mkosi/machine.py

```diff
@@ -122,9 +143,55 @@ def run(self, commands: Sequence[str], timeout: int = 900, check: bool = True) -
 
         return process
 
-    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+    def _kill(self, *args: Any, **kwargs: Any) -> None:
```

Comment:
When you make kill() a public method like I suggested below, I'd also remove the `args` and `kwargs` arguments to make the method easier to use. We can switch things around and have kill() call `__exit__()` instead of `__exit__()` calling kill(). kill() can then call `__exit__()` with arguments `(None, None, None)` and we can then remove `args` and `kwargs` from kill()

Reply:
I'm unable to remove `args` and `kwargs` when running QEMU:

```
>               _fix_exception_context(new_exc_details[1], exc_details[1])
E               IndexError: tuple index out of range
```

Couldn't find a way to overcome this

Reply:
What I'm proposing is the following:

```
def kill(self):
    self.__exit__(None, None, None)

def __exit__(self, *args, **kwargs):
    if self._serial:
        self._serial.kill(signal.SIGTERM)
        self.exit_code = self._serial.wait()
        self._serial = None
    self.stack.__exit__(*args, **kwargs)
```

Reply:
oh, I understood ((None, None, None)), this works